cpp/src/arrow/compute/kernels/scalar_nested.cc

```diff
@@ -428,6 +429,276 @@ const FunctionDoc make_struct_doc{"Wrap Arrays into a StructArray",
+  const auto& options = OptionsWrapper<MapArrayLookupOptions>::Get(ctx);
+  std::shared_ptr<DataType> type = descrs.front().type;
+  std::shared_ptr<DataType> item_type = checked_cast<const MapType&>(*type).item_type();
+  std::shared_ptr<DataType> key_type = checked_cast<const MapType&>(*type).key_type();
+
+  if (!options.query_key) {
+    return Status::TypeError("map_array_lookup: query_key can't be empty.");
+  } else if (!options.query_key->is_valid) {
+    return Status::TypeError("map_array_lookup: query_key can't be null.");
+  } else if (!options.query_key->type || !options.query_key->type->Equals(key_type)) {
+    return Status::TypeError(
+        "map_array_lookup: query_key type and MapArray key_type don't match. Expected "
+        "type: ",
+        *key_type, ", but got type: ", *options.query_key->type);
+  }
+
+  if (options.occurrence == MapArrayLookupOptions::Occurrence::ALL) {
+    return ValueDescr(list(item_type), descrs.front().shape);
+  } else { /* occurrence == FIRST || LAST */
+    return ValueDescr(item_type, descrs.front().shape);
+  }
+}
+
+struct ResolveMapArrayLookup {
+  KernelContext* ctx;
+  const ExecBatch& batch;
+  Datum* out;
+
+  template <typename KeyType>
+  Status Execute() {
+    if (batch[0].kind() == Datum::SCALAR) {
+      return MapArrayLookupFunctor<KeyType>::ExecMapScalar(ctx, batch, out);
+    }
+    return MapArrayLookupFunctor<KeyType>::ExecMapArray(ctx, batch, out);
+  }
+
+  template <typename KeyType>
+  enable_if_physical_integer<KeyType, Status> Visit(const KeyType& type) {
+    return Execute<KeyType>();
+  }
+
+  template <typename KeyType>
+  enable_if_decimal<KeyType, Status> Visit(const KeyType& type) {
+    return Execute<KeyType>();
+  }
+
+  template <typename KeyType>
+  enable_if_base_binary<KeyType, Status> Visit(const KeyType& type) {
+    return Execute<KeyType>();
+  }
+
+  template <typename KeyType>
+  enable_if_boolean<KeyType, Status> Visit(const KeyType& type) {
+    return Execute<KeyType>();
+  }
+
+  template <typename KeyType>
+  enable_if_same<KeyType, FixedSizeBinaryType, Status> Visit(const KeyType& key) {
+    return Execute<KeyType>();
+  }
+
+  Status Visit(const MonthDayNanoIntervalType& key) {
+    return Execute<MonthDayNanoIntervalType>();
+  }
+
+  Status Visit(const DataType& type) {
+    return Status::TypeError("Got unsupported type: ", type.ToString());
+  }
+
+  static Status Exec(KernelContext* ctx, const ExecBatch& batch, Datum* out) {
+    ResolveMapArrayLookup visitor{ctx, batch, out};
+    return VisitTypeInline(*checked_cast<const MapType&>(*batch[0].type()).key_type(),
+                           &visitor);
+  }
+};
+
+void AddMapArrayLookupKernels(ScalarFunction* func) {
+  for (const auto shape : {ValueDescr::ARRAY, ValueDescr::SCALAR}) {
+    ScalarKernel kernel(
+        {InputType(Type::MAP, shape)}, OutputType(ResolveMapArrayLookupType),
+        ResolveMapArrayLookup::Exec, OptionsWrapper<MapArrayLookupOptions>::Init);
```

Comment:
Since the Exec is the same regardless of the shape, we can remove the loop and just register one kernel for `InputType(Type::MAP)`